lib/commands/performance.js

```diff
@@ -13,31 +16,179 @@ const PERF_RECORD_FEAT_NAME = 'perf_record';
   'referencing https://github.com/appium/appium/blob/master/docs/en/writing-running-appium/security.md for more details.';
-const RECORDERS_CACHE = {};
 const DEFAULT_TIMEOUT_MS = 5 * 60 * 1000;
 const STOP_TIMEOUT_MS = 3 * 60 * 1000;
-const PROFILING_TIMEOUT_MS = 15 * 1000;
+const STARTUP_TIMEOUT_MS = 15 * 1000;
 const DEFAULT_PROFILE_NAME = 'Activity Monitor';
-const DEFAULT_EXT = '.trace';
+const DEFAULT_EXT = 'trace';
+const DEFAULT_PID = 'current';
+const INSTRUMENTS_BINARY = 'instruments';
 
 
-async function finishPerfRecord (proc, stopGracefully = true) {
-  if (!proc.isRunning) {
-    return;
-  }
-  if (stopGracefully) {
-    log.debug(`Sending SIGINT to the running instruments process`);
-    return await proc.stop('SIGINT', STOP_TIMEOUT_MS);
+async function requireInstruments () {
+  try {
+    return await fs.which(INSTRUMENTS_BINARY);
+  } catch (e) {
+    log.errorAndThrow(`${INSTRUMENTS_BINARY} has not been found in PATH. ` +
+      `Please make sure Xcode development tools are installed`);
   }
-  log.debug(`Sending SIGTERM to the running instruments process`);
-  await proc.stop();
 }
 
-async function uploadTrace (localFile, remotePath = null, uploadOptions = {}) {
-  try {
-    return await encodeBase64OrUpload(localFile, remotePath, uploadOptions);
-  } finally {
-    await fs.rimraf(localFile);
+
+class PerfRecorder {
+  constructor (reportPath, udid, opts = {}) {
+    this._process = null;
+    this._reportPath = reportPath;
+    this._zippedReportPath = '';
+    this._timeout = (opts.timeout && opts.timeout > 0) ? opts.timeout : DEFAULT_TIMEOUT_MS;
+    this._profileName = opts.profileName || DEFAULT_PROFILE_NAME;
+    this._pid = opts.pid;
+    this._udid = udid;
+    this._logger = logger.getLogger(
+      `${_.truncate(this.profileName, {length: 10})}@${this.udid.substring(0, 8)}`);
+    this._archivePromise = null;
+  }
+
+  get profileName () {
+    return this._profileName;
+  }
+
+  async getOriginalReportPath () {
+    return (await fs.exists(this._reportPath)) ? this._reportPath : '';
+  }
+
+  async getZippedReportPath () {
+    if (await fs.exists(this._zippedReportPath)) {
+      return this._zippedReportPath;
+    }
+    const originalReportPath = await this.getOriginalReportPath();
+    if (!originalReportPath) {
+      return '';
+    }
+    const zippedReportPath = originalReportPath.replace(`.${DEFAULT_EXT}`, '.zip');
+    // This is to prevent possible race conditions, because the archive operation
+    // could be pretty time-intensive
+    if (!this._archivePromise) {
+      this._archivePromise = zip.toArchive(zippedReportPath, {
+        cwd: originalReportPath,
+      });
+    }
+    await this._archivePromise;
+    this._zippedReportPath = zippedReportPath;
+    return this._zippedReportPath;
+  }
+
+  isRunning () {
+    return !!(this._process?.isRunning);
+  }
+
+  async _enforceTermination () {
+    if (this._process && this.isRunning()) {
```

Comment:
`this._process` is not necessary here.

Reply:
I've added it deliberately for better readability.

Reply:
Not sure I understand how that's more readable. But ok.